File: src/main/kotlin/com/raychenon/kotlin/leetcode/MinimumPathSum.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 21/4/2020
 * https://leetcode.com/problems/minimum-path-sum/
 */
object MinimumPathSum {

    /**
     * Time complexity: O(W * H) , width * height of matrix
     * Space complexity: O(W * H)
     */
    fun minPathSum(grid: Array<IntArray>): Int {
        val W = grid.size
        val H = grid[0].size
        val dp = Array(W) { IntArray(H) }

        for (i in 0 until W) {
            for (j in 0 until H) {
                dp[i][j] = grid[i][j]
                if (i > 0 && j > 0) {
                    dp[i][j] += Math.min(dp[i - 1][j], dp[i][j - 1])
                } else if (i > 0) {
                    dp[i][j] += dp[i - 1][j]
                } else if (j > 0) {
                    dp[i][j] += dp[i][j - 1]
                }
            }
        }
        return dp[W - 1][H - 1]
    }

}

File: python3/algorithms/leetcode/minimum_path_sum.py
"""
https://leetcode.com/problems/minimum-path-sum/
"""

# O(n) space
def minPathSumNSpace(grid) -> int:
    if not grid:
        return
    r,c = len(grid), len(grid[0])
    cur = [0] * c
    cur[0] = grid[0][0]
    for i in range(1,c):
        cur[i] = cur[i-1] + grid[0][i]
    for j in range(1,r):
        cur[0] += grid[i][0]
        for j in range(1, c):
    cur[j] = min(cur[j-1], cur[j]) + grid[i][j]
    return cur[-1]


def assertMethods(expected, grid):
    assert minPathSumNSpace(grid) == expected


if __name__ == "__main__":
    assertMethods(7,[
        [1,3,1],
        [1,5,1],
        [4,2,1]
    ])
